### PR TITLE
Fix alignment of scratchBuffer in HUF_compressWeights()

### PR DESCRIPTION
**Pull request by terrelln**
The scratch buffer must be 4-byte aligned. This causes test failures in 32-bit systems, where the stack isn't aligned.

Fixes Issue #2428.

### CODE REVIEW

**Comment by terrelln** [comment]
`make check` in 32-bit mode is still failing, but this new failure seems like a test problem, where we're trying to run a test that allocates too much memory.

```
===>  zstd long distance matching with optimal parser compressed size tests
zstd: error 11 : Allocation error : not enough memory
```

**Comment by Cyan4973** [comment]
Great fix @terrelln !

Should we also add a test, able to trigger that issue (when unfixed) ?

**Comment by Cyan4973** [comment]
> make check in 32-bit mode is still failing, but this new failure seems like a test problem, where we're trying to run a test that allocates too much memory.

Indeed,
maybe we should find a way to reduce memory requirement for this test.

**Comment by terrelln** [comment]
Yes, we should run `make check` in 32-bit mode without any sanitizers. I don't know any other way to trigger the bug. But, it is failing later on, so I can't add it yet.

I'm going to be away for an hour or two now, so I can't look into it immediately. When I get back I will look into it, unless someone beats me to the punch.

**Comment by Cyan4973** [comment]
I'm guessing `zstd -22 --ultra` is using way too much resources, especially when combined with `stdin` streaming

File: lib/compress/huf_compress.c [hunk 1]
```diff
@@ -69,7 +69,7 @@ static size_t HUF_compressWeights (void* dst, size_t dstSize, const void* weight
     U32 tableLog = MAX_FSE_TABLELOG_FOR_HUFF_HEADER;
 
     FSE_CTable CTable[FSE_CTABLE_SIZE_U32(MAX_FSE_TABLELOG_FOR_HUFF_HEADER, HUF_TABLELOG_MAX)];
-    BYTE scratchBuffer[FSE_BUILD_CTABLE_WORKSPACE_SIZE(HUF_TABLELOG_MAX, MAX_FSE_TABLELOG_FOR_HUFF_HEADER)];
+    U32 scratchBuffer[FSE_BUILD_CTABLE_WORKSPACE_SIZE_U32(HUF_TABLELOG_MAX, MAX_FSE_TABLELOG_FOR_HUFF_HEADER)];
```
Comment (Cyan4973) [inline lib/compress/huf_compress.c:72]: minor question:

I'm struggling to understand how this could be an issue _only_ for 32-bit systems ...

Comment (terrelln) [reply lib/compress/huf_compress.c:72]: I believe that either compilers or the x86-64 ABI always align all stack entries to 8 or 16 bytes. But 32-bit systems wont.

**Comment by Cyan4973** [comment]
Redirecting the PR into `check32`, where a 32-bit runtime test is added, and currently fails (as expected).
The fix in this PR should make the new test pass.

File: lib/compress/huf_compress.c (hunk 1 above)
Comment (Cyan4973) [reply lib/compress/huf_compress.c:72]: That's a good guess.
It's in line with observations that the bug disappears on enabling sanitizers.

**Review by Cyan4973 — APPROVED** [review]

File: lib/compress/huf_compress.c (hunk 1 above)
Comment (bmwiedemann) [reply lib/compress/huf_compress.c:72]: s390x also failed and is a 64bit arch. So probably more related to the alignment standards applied by compilers for the arch.

**Pull request merged**